pkg/render/typha.go

```diff
@@ -607,17 +607,43 @@ func (c *typhaComponent) typhaPodSecurityPolicy() *policyv1beta1.PodSecurityPoli
 	return psp
 }
 
-// toAffinity converts a typha affinity to a k8s affinity.
-func toAffinity(t *operator.TyphaAffinity) *v1.Affinity {
-	if t == nil {
-		return nil
-	}
-	if t.NodeAffinity == nil {
-		return nil
+// affinity sets the affinity on typha, accounting for the kubernetes-provider and user-specified values.
+func (c *typhaComponent) affinity() (aff *v1.Affinity) {
+	// in AKS, there is a feature called 'virtual-nodes' which represent azure's container service as a node in the kubernetes cluster.
+	// virtual-nodes have many limitations, namely it's unable to run hostNetworked pods. virtual-kubelets are tainted to prevent pods from running on them,
+	// but typha tolerates all taints and will run there.
+	// as such, we add a required anti-affinity for virtual-nodes if running on azure
+	if c.installation.KubernetesProvider == operator.ProviderAKS {
+		aff = &v1.Affinity{
+			NodeAffinity: &v1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+					NodeSelectorTerms: []v1.NodeSelectorTerm{{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{
+								Key:      "type",
+								Operator: corev1.NodeSelectorOpNotIn,
+								Values:   []string{"virtual-node"},
+							},
+							{
+								Key:      "kubernetes.azure.com/cluster",
```

Comment:
I'm second guessing including this in the match here. If this Provider was set but nodes did not have `kubernetes.azure.com/cluster` then there would be no nodes matched. That doesn't seem very helpful. I feel like we'd want the logic to be
```
if key kubernetes.azure.com/cluster key exists and key type has value virtual-node
then do not match
```
but this is actually requiring key kubernetes.azure.com/cluster to match.

Reply:
I don't think we can achieve the logic that we would want so maybe this should just be the `key type is not in virtual-node`

Reply:
Actually I think we might be able to get the behavior we'd want with
```
NodeSelectorTerms: []v1.NodeSelectorTerm{{
    MatchExpressions: []v1.NodeSelectorRequirement{
        {
            Key:      "kubernetes.azure.com/cluster",
            Operator: v1.NodeSelectorOpDoesNotExist,
        },
    },
    MatchExpressions: []v1.NodeSelectorRequirement{
        {
            Key:      "type",
            Operator: corev1.NodeSelectorOpNotIn,
            Values:   []string{"virtual-node"},
        },
        {
            Key:      "kubernetes.azure.com/cluster",
            Operator: v1.NodeSelectorOpExists,
        },  
    },  
}} 
```
That way any `kubernetes.azure.com/cluster` nodes that do not have type `virtual-node` can be selected or any nodes that do not have `kuberentes.azure.com/cluster` can be selected.

Reply:
We got confirmation that all AKS nodes will have the `kubernetes.azure.com/cluster` label so this should be good to go.